File: packages/sdk-python/src/melaya/backtest.py
"""Backtest API — run strategies against historical data on the Rust engine.

Start a single run or a parameter sweep (grid / random), poll the job, then
pull metrics, the equity curve, and the trade list. All backtests run natively
on Melaya's in-house engine.
"""
from __future__ import annotations

from typing import Any, Callable, Dict, List, Optional

_Request = Callable[..., Any]


class BacktestAPI:
    def __init__(self, request: _Request) -> None:
        self._request = request

    def start(self, body: Dict[str, Any]) -> Dict[str, Any]:
        """Start a backtest. Returns the job id(s); poll with ``job()``.

        Pass ``mode='grid_sweep'`` / ``'random_sweep'`` with ``paramRanges`` to
        fan out a parameter search; omit ``mode`` for a single run.
        """
        return self._request("POST", "/api/v1/private/backtest/start", json=body)

    def job(self, job_id: str) -> Dict[str, Any]:
        """Job status + progress."""
        return self._request("GET", f"/api/v1/private/backtest/jobs/{job_id}")

    def results(self, job_id: str) -> Dict[str, Any]:
        """Metrics, equity curve, and OHLCV for a completed job."""
        return self._request("GET", f"/api/v1/private/backtest/results/{job_id}")["result"]

    def trades(self, job_id: str, *, limit: Optional[int] = None, offset: Optional[int] = None) -> List[Any]:
        """The trade list for a completed job (default 500, max 5000 per call)."""
        return self._request("GET", f"/api/v1/private/backtest/trades/{job_id}",
                             params={"limit": limit, "offset": offset})["trades"]

    def sweep(self, parent_id: str, *, objective: Optional[str] = None,
              limit: Optional[int] = None) -> Dict[str, Any]:
        """Ranked children of a sweep parent (default objective: sharpe DESC)."""
        return self._request("GET", f"/api/v1/private/backtest/sweep/{parent_id}",
                             params={"objective": objective, "limit": limit})

    def list(self, *, limit: Optional[int] = None, offset: Optional[int] = None) -> List[Any]:
        """Your backtest jobs, newest first."""
        return self._request("GET", "/api/v1/private/backtest",
                             params={"limit": limit, "offset": offset})["data"]["jobs"]

    def favorites(self, *, limit: Optional[int] = None, offset: Optional[int] = None) -> List[Any]:
        """Your favorited backtest jobs (Forge tier and above)."""
        return self._request("GET", "/api/v1/private/backtest/favorites",
                             params={"limit": limit, "offset": offset})["data"]["jobs"]

    def funding_range(self, *, exchange: str, symbol: str) -> Optional[int]:
        """Earliest funding-rate timestamp available for exchange+symbol (ms, or None)."""
        return self._request("GET", "/api/v1/private/backtest/funding-range",
                             params={"exchange": exchange, "symbol": symbol})["earliest_ms"]

    def cancel(self, job_id: str) -> Dict[str, Any]:
        """Cancel an in-flight job."""
        return self._request("POST", f"/api/v1/private/backtest/{job_id}/cancel")

    def delete(self, job_id: str) -> Dict[str, Any]:
        """Soft-delete a single job."""
        return self._request("DELETE", f"/api/v1/private/backtest/{job_id}")

    def delete_all(self) -> Dict[str, Any]:
        """Soft-delete every non-favorited job. Returns the count deleted."""
        return self._request("DELETE", "/api/v1/private/backtest")

File: packages/sdk-rust/src/backtest.rs
use std::collections::HashMap;

use serde_json::{json, Value};

use crate::client::HttpClient;
use crate::error::Result;

/// Backtest API — run strategies against historical data on the Rust engine.
pub struct BacktestAPI {
    http: HttpClient,
}

impl BacktestAPI {
    pub(crate) fn new(http: HttpClient) -> Self {
        Self { http }
    }

    /// Start a backtest. Returns the job id(s) — poll with [`BacktestAPI::job`].
    pub async fn start(&self, body: &Value) -> Result<Value> {
        self.http.post("/api/v1/private/backtest/start", body).await
    }

    /// Job status + progress.
    pub async fn job(&self, job_id: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .get(&format!("/api/v1/private/backtest/jobs/{job_id}"), &q)
            .await
    }

    /// Metrics, equity curve, and OHLCV for a completed job.
    pub async fn results(&self, job_id: &str) -> Result<Value> {
        let q = HashMap::new();
        let r = self
            .http
            .get(&format!("/api/v1/private/backtest/results/{job_id}"), &q)
            .await?;
        Ok(r["result"].clone())
    }

    /// The trade list for a completed job.
    pub async fn trades(
        &self,
        job_id: &str,
        limit: Option<u32>,
        offset: Option<u32>,
    ) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("limit", limit.map(|v| v.to_string()));
        q.insert("offset", offset.map(|v| v.to_string()));
        let r = self
            .http
            .get(&format!("/api/v1/private/backtest/trades/{job_id}"), &q)
            .await?;
        Ok(r["trades"].clone())
    }

    /// Ranked children of a sweep parent.
    pub async fn sweep(
        &self,
        parent_id: &str,
        objective: Option<&str>,
        limit: Option<u32>,
    ) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("objective", objective.map(str::to_owned));
        q.insert("limit", limit.map(|v| v.to_string()));
        self.http
            .get(
                &format!("/api/v1/private/backtest/sweep/{parent_id}"),
                &q,
            )
            .await
    }

    /// Your backtest jobs, newest first.
    pub async fn list(&self, limit: Option<u32>, offset: Option<u32>) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("limit", limit.map(|v| v.to_string()));
        q.insert("offset", offset.map(|v| v.to_string()));
        let r = self.http.get("/api/v1/private/backtest", &q).await?;
        Ok(r["data"]["jobs"].clone())
    }

    /// Your favorited backtest jobs (Forge tier and above).
    pub async fn favorites(&self, limit: Option<u32>, offset: Option<u32>) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("limit", limit.map(|v| v.to_string()));
        q.insert("offset", offset.map(|v| v.to_string()));
        let r = self
            .http
            .get("/api/v1/private/backtest/favorites", &q)
            .await?;
        Ok(r["data"]["jobs"].clone())
    }

    /// Earliest funding-rate timestamp available for an exchange+symbol (ms, or null).
    pub async fn funding_range(&self, exchange: &str, symbol: &str) -> Result<Value> {
        let mut q: HashMap<&str, Option<String>> = HashMap::new();
        q.insert("exchange", Some(exchange.to_owned()));
        q.insert("symbol", Some(symbol.to_owned()));
        let r = self
            .http
            .get("/api/v1/private/backtest/funding-range", &q)
            .await?;
        Ok(r["earliest_ms"].clone())
    }

    /// Cancel an in-flight job.
    pub async fn cancel(&self, job_id: &str) -> Result<Value> {
        self.http
            .post(
                &format!("/api/v1/private/backtest/{job_id}/cancel"),
                &json!({}),
            )
            .await
    }

    /// Soft-delete a single job.
    pub async fn delete(&self, job_id: &str) -> Result<Value> {
        let q = HashMap::new();
        self.http
            .delete(&format!("/api/v1/private/backtest/{job_id}"), &q)
            .await
    }

    /// Soft-delete every non-favorited job.
    pub async fn delete_all(&self) -> Result<Value> {
        let q = HashMap::new();
        self.http.delete("/api/v1/private/backtest", &q).await
    }
}
